src/sentry/models/notificationmessage.py

```diff
@@ -0,0 +1,97 @@
+from django.db.models import DateTimeField, IntegerField, Q
+from django.db.models.constraints import CheckConstraint, UniqueConstraint
+from django.utils import timezone
+
+from sentry.backup.scopes import RelocationScope
+from sentry.db.models import (
+    CharField,
+    FlexibleForeignKey,
+    JSONField,
+    Model,
+    region_silo_only_model,
+    sane_repr,
+)
+
+
+@region_silo_only_model
+class NotificationMessage(Model):
+    """
+    Data model represents the aggregate for an entire notification message.
+    A notification message must always have the resulting identifier that was sent back if there was no error.
+    When there is an error, there will be no identifier to leverage or store.
+
+    If a notification message has an error, the resulting error code and details will be saved.
+    This information will then be leveraged to show the user which notifications are consistently failing.
+    An example of this would be if a Slack channel no longer exists.
+
+    The data model hold the singular gateway for both Metric, and Issue, Alerts.
+    Following the specific data relationship, you should be able to find the specific rule, organization, project, and
+    incident or group event that this notification message was enacted.
+
+    If any notification message is in relation to another notification message, you should be able to find the original
+    notification message through the parent notification message relation.
+    """
+
+    __relocation_scope__ = RelocationScope.Excluded
+
+    # Related information regarding failed notifications.
+    # Leveraged to help give the user visibility into notifications that are consistently failing.
+    error_details = JSONField(null=True)
+    error_code = IntegerField(null=True, db_index=True)
+
+    # Resulting identifier from the vendor that can be leveraged for future interaction with the notification.
+    message_identifier = CharField(null=True)
+    # Reference to another notification if we choose to modify the original message or reply to it (like start a thread)
+    parent_notification_message = FlexibleForeignKey("self", null=True)
+
+    # Related information regarding Alert Rules (Metric Alerts)
+    incident = FlexibleForeignKey("sentry.Incident", null=True)
+    trigger_action = FlexibleForeignKey("sentry.AlertRuleTriggerAction", null=True)
```

Comment:
You might need to set up something in deletions for these rows that are associated with `AlertRuleTriggerAction`. You usually need to do this when there's a chance that when the related row is deleted, there will be a lot of deleted rows from this table.

So in this case, a `trigger_action` could have an unbound number of `NotificationMessage` rows associated with it. If we rely on cascade delete here, this could be a pretty large transaction that would end up timing out.

An example of this for issue alerts where we delete `RuleFireHistory` rows is here.
https://github.com/getsentry/sentry/blob/714fabf11fc03a591f9ddc98dd5ec6561502ea3e/src/sentry/deletions/defaults/rule.py#L4

Just noting that we probably don't need this for `RuleFireHistory` or `Incident` since they inherently won't have a lot of `NotificationMessage` associated with them.

Reply:
Ohhh, let me look into this, thank you for calling that out! I can open another PR so the change is easier to see. I won't start using the table until I have this deletion setup

Reply:
Following up here: https://github.com/getsentry/sentry/pull/65287